### PR TITLE
ConfigurableAgeOffFilter revised caching implementation

### PR DESCRIPTION
**Pull request by dtspence**
Revised cache loader implementation that addresses issues experienced with ConfigurableAgeOffFilter when performing a cache refresh. The logic implements a revised cache loading shifts the cache refresh checks into the loader implementation.

### CODE REVIEW

File: warehouse/age-off/src/main/java/datawave/iterators/filter/ConfigurableAgeOffFilter.java [hunk 1]
```diff
@@ -435,14 +424,15 @@ private boolean shouldDisableForNonUserCompaction(Map<String,String> options, It
...
      */
     private void initFilterRules() throws IllegalArgumentException, IOException {
-        // filename
         if (null == ruleCache) {
             synchronized (ConfigurableAgeOffFilter.class) {
                 if (null == ruleCache) {
                     UPDATE_INTERVAL_MS = getLongProperty(UPDATE_INTERVAL_MS_PROP, DEFAULT_UPDATE_INTERVAL_MS); // 5 ms
                     EXPIRATION_INTERVAL_MS = getLongProperty(EXPIRATION_INTERVAL_MS_PROP, DEFAULT_EXPIRATION_INTERVAL_MS); // 1 hour
+                    log.debug("Configured refresh interval: " + UPDATE_INTERVAL_MS);
+                    log.debug("Configured expiration interval: " + EXPIRATION_INTERVAL_MS);
```
Comment (apmoriarty) [inline warehouse/age-off/src/main/java/datawave/iterators/filter/ConfigurableAgeOffFilter.java]:
```suggestion
                    log.debug("Configured refresh interval (ms): " + UPDATE_INTERVAL_MS);
                    log.debug("Configured expiration interval (ms): " + EXPIRATION_INTERVAL_MS);
```

File: warehouse/age-off/src/main/java/datawave/ingest/util/cache/watch/FileRuleWatcher.java [hunk 2]
```diff
@@ -89,7 +88,7 @@ public FileRuleWatcher(Path filePath, long configuredDiff, IteratorEnvironment i
     @Override
     protected Collection<FilterRule> loadContents(InputStream in) throws IOException {
         try {
-            AgeOffRuleLoader ruleLoader = new AgeOffRuleLoader(new FileWatcherDependencyProvider());
+            AgeOffRuleLoader ruleLoader = new AgeOffRuleLoader(new FileLoaderDependencyProvider(fs, filePath, iterEnv));
             return ruleLoader.load(in);
         } catch (Exception ex) {
             log.error("uh oh: " + ex);
```
Comment (avgAGB) [inline warehouse/age-off/src/main/java/datawave/ingest/util/cache/watch/FileRuleWatcher.java]: maybe the message here could be tweaked a little, not a big deal though

**Review by apmoriarty — DISMISSED** [review]

File: warehouse/age-off/src/main/java/datawave/ingest/util/cache/watch/FileRuleWatcher.java (hunk 2 above)
Comment (dtspence) [reply warehouse/age-off/src/main/java/datawave/ingest/util/cache/watch/FileRuleWatcher.java]: I updated a few log message locations when an exception occurs.

**Review by apmoriarty — DISMISSED** [review]

File: warehouse/age-off/src/main/java/datawave/ingest/util/cache/watch/AgeOffRuleLoader.java [hunk 3]
```diff
@@ -134,7 +134,7 @@ protected List<RuleConfig> loadRuleConfigs(InputStream in) throws IOException {
 
             // @formatter:on
         } catch (Exception ex) {
```
Comment (FineAndDandy) [inline warehouse/age-off/src/main/java/datawave/ingest/util/cache/watch/AgeOffRuleLoader.java]: can this exception be updated to be more specific?

File: warehouse/age-off/src/main/java/datawave/ingest/util/cache/watch/FileRuleCacheValue.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        return ruleRef.deepCopy(scanStart, iterEnv);
+    }
+
+    @VisibleForTesting
+    Collection<FilterRule> loadFilterRules(IteratorEnvironment iterEnv) throws IOException {
+        AgeOffRuleLoader ruleLoader = new AgeOffRuleLoader(new FileLoaderDependencyProvider(fs, filePath, iterEnv));
+        Collection<FilterRule> rulesBase;
+        try (InputStream in = fs.open(filePath)) {
+            rulesBase = ruleLoader.load(in);
+        } catch (Exception e) {
```
Comment (FineAndDandy) [inline warehouse/age-off/src/main/java/datawave/ingest/util/cache/watch/FileRuleCacheValue.java]: update to be a more specific catch

File: warehouse/age-off/src/main/java/datawave/ingest/util/cache/watch/FileRuleWatcher.java [hunk 5]
```diff
@@ -89,37 +88,13 @@ public FileRuleWatcher(Path filePath, long configuredDiff, IteratorEnvironment i
     @Override
     protected Collection<FilterRule> loadContents(InputStream in) throws IOException {
         try {
-            AgeOffRuleLoader ruleLoader = new AgeOffRuleLoader(new FileWatcherDependencyProvider());
+            AgeOffRuleLoader ruleLoader = new AgeOffRuleLoader(new FileLoaderDependencyProvider(fs, filePath, iterEnv));
             return ruleLoader.load(in);
         } catch (Exception ex) {
```
Comment (FineAndDandy) [inline warehouse/age-off/src/main/java/datawave/ingest/util/cache/watch/FileRuleWatcher.java]: likewise, update to be more specific

**Review by FineAndDandy — CHANGES_REQUESTED** [review]
just some exception handling requests

File: warehouse/age-off/src/main/java/datawave/ingest/util/cache/watch/AgeOffRuleLoader.java (hunk 3 above)
Comment (dtspence) [reply warehouse/age-off/src/main/java/datawave/ingest/util/cache/watch/AgeOffRuleLoader.java]: Updated the exception to be more specific based on surrounding code. Introduced an argument exception based the filter class being missing. Additionally moved a couple of exceptions on the method signature to be rethrown as `IOException`, based on surrounding code.

File: warehouse/age-off/src/main/java/datawave/ingest/util/cache/watch/FileRuleCacheValue.java (hunk 4 above)
Comment (dtspence) [reply warehouse/age-off/src/main/java/datawave/ingest/util/cache/watch/FileRuleCacheValue.java]: Removed general exception catch - `IOException` remained on the method as a checked exception.

File: warehouse/age-off/src/main/java/datawave/ingest/util/cache/watch/FileRuleWatcher.java (hunk 5 above)
Comment (dtspence) [reply warehouse/age-off/src/main/java/datawave/ingest/util/cache/watch/FileRuleWatcher.java]: Removed general exception catch - `IOException` remained on the method as a checked exception.

**Review by FineAndDandy — APPROVED** [review]

**Review by apmoriarty — APPROVED** [review]

**Pull request merged**